Add Nvidia Jetson Series Support

pjueon opened the pull request:
related to #22 

- Tested on Jetson Nano & JJY based watch (Citizen REGUNO KL8-911-71), with both JJY40 and JJY60.
- Can choose the platform by the cmake option `PLATFORM`. (ex> `-DPLATFORM=jetson`)
  - default value is `rpi` 
- Isolate the platform-specific information from the main logic so that the new other platform supports could be easily added in the future.

Main differences from Raspberry Pi platform:
- [JetsonGPIO](https://github.com/pjueon/JetsonGPIO) library is used to control the hardware. 
- The PWM pin is used as frequency output and it works with switching the signal (on-off keying) instead of attenuating. 
    - Same approaches as other similar projects: [micooke/wwvb_jjy](https://github.com/micooke/wwvb_jjy), [nsicko42/jjy-emulator-using-raspberry-pi](https://github.com/nsicko42/jjy-emulator-using-raspberry-pi)

hzeller commented:
Very nice, thanks! I'll have a first look at the PR.

I think we can't leave out the attenuation feature as some time signals require it; for instance DCF77 has the carrier always on, but [specifically calls for attenuation](https://en.wikipedia.org/wiki/DCF77#Amplitude_modulation). So for completeness and feature parity with the Raspberry Pi hardware, we need both `LOW` and `OFF`.

hzeller commented on .gitignore:
```diff
@@ -1,3 +1,46 @@
+
+# End of https://www.toptal.com/developers/gitignore/api/linux,visualstudiocode
+
+build/*
```
This does look like a lot of IDE gnerated stuff in here that should not necessarily make it upstream; every developer with specific IDEs should keep their .gitignore with whatever is needed in theri context.

So here, we can just simplify the additions to `.gitignore` to`build/*` to accomodate the cmake-project.

hzeller commented on include/carrier-power.h:
```diff
@@ -0,0 +1,25 @@
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef ENUMS
```
I'd call this include-guard after the file, so `CAFFIER_POWER_H`

hzeller commented on include/rpi/hardware-control-implementation.h:
```diff
@@ -15,20 +15,21 @@
 
-#ifndef RPI_GPIO_H
-#define RPI_GPIO_H
+#ifndef RPI_CONTROL_H
```
simiilar here: let's derive the include-guard from the filename `RPI_HARDWARE_CONTROL_IMPLEMENTATION_H`

hzeller commented on include/rpi/hardware-control-implementation.h:
```diff
@@ -50,15 +51,18 @@ class GPIO {
-#endif  // RPI_GPIO_H
+using GPIO = HardwareControl::Implementation;
+
+#endif
```
Please make sure to have a newline at the end of each file (the .editorconfig file also will configure that)

hzeller commented on include/rpi/hardware-control-implementation.h:
```diff
@@ -50,15 +51,18 @@ class GPIO {
   double StartClock(double frequency_hertz);
   void StopClock();
-
+  
```
Don't leave trailing whitespaces in lines.
(I have just added a [`.editorconfig`](https://github.com/hzeller/txtempus/blob/master/.editorconfig) file so that it is easier to automatically do, as most editors can use that file).

hzeller commented on README.md:
```diff
@@ -100,13 +108,55 @@ adding an amplifier. *Only go in this direction after familiarizing yourself
+
+We use hardware PWM pin as a frequency output pin. Unlike the Raspberry Pi, we need only one output pin 
+because we modulate the signal by switching (on-off keying) instead of attenuating.
+And we use only one register (4.7kΩ) for the Jetson.
```
s/register/resistor/

hzeller commented on README.md:
```diff
@@ -100,13 +108,55 @@ adding an amplifier. *Only go in this direction after familiarizing yourself
+Here's the full schematic of the external hardware for Jetson Series:
+Schematic                      | Real world
+-------------------------------|------------------------------
+![](img/schematic-jetson.jpg)   |![](img/jetson-nano.jpg)
```
Not a biggie, but the 'real world picture' would be even better if the viewer could better see where the pins are connected on the Pin header. So more taking a picture from the right side so that you can see onto the full pin-header and see which pins are connected while still having the watch in the foreground.

hzeller commented on README.md:
```diff
@@ -100,13 +108,55 @@ adding an amplifier. *Only go in this direction after familiarizing yourself
+On Jetson, the external hardware setup is slightly different from Raspberry Pi. 
+
+We use hardware PWM pin as a frequency output pin. Unlike the Raspberry Pi, we need only one output pin 
+because we modulate the signal by switching (on-off keying) instead of attenuating.
```
Note, some of the signals require attenuation to be in spec. So if you want to skip implement attenuation in the first step, maybe point out that it might not work in all circumstances.

hzeller reviewed (commented):
Very nice!
The separation of the hardware interfaces is nicely done, I like the separation into directories.

I just have a few smallish remarks before we're ready to merge.

pjueon commented:
Thanks for the detailed feedback!
I'll edit it when I get home.

hzeller commented:
If the voltage is 1.8V, this indidcates that the GPIO ports are connected to the internal core voltage of the CPU and they will have issue to drive anything and it will probably dangerous to even connecting the coil, as these tiny mosfets in the output stage will not be able to move the current.

I strongly suggest to connect the GPIO pins via a 1.8V -> 3.3V level shifter to (a) relieve the GPIO pins (b) have enough drive strength for the external components such as coil and pulldown.

pjueon commented:
> If the voltage is 1.8V, this indidcates that the GPIO ports are connected to the internal core voltage of the CPU and they will have issue to drive anything and it will probably dangerous to even connecting the coil, as these tiny mosfets in the output stage will not be able to move the current.
> 
> I strongly suggest to connect the GPIO pins via a 1.8V -> 3.3V level shifter to (a) relieve the GPIO pins (b) have enough drive strength for the external components such as coil and pulldown.

I have too little knowledge about HW stuff, so I can't understand what you're suggesting. Could you explain with a picture please? 

And when I asked this problem to Nvidia developers forum, someone linked this pdf. Maybe it's related. 
[customizing_the_jetson_nano_40-pin_expansion_header_v1.2.pdf.pdf](https://github.com/hzeller/txtempus/files/8068264/customizing_the_jetson_nano_40-pin_expansion_header_v1.2.pdf.pdf)

It's about the level shifters in Jetson Nano, but I cannot understand it. I don't even know what "level shifters", "buffers", "driving"  mean...

icb- commented:
A level shifter translates a signal with one voltage (like 1.8V) to a signal with a different voltage (like 3.3V). A buffer takes a signal and repeats it on its output. Buffers can often do level shifting, depending on their input thresholds.

In the circuit below, you have two tri-state buffers doing level shifting. Tri-state means the output can be driven (set to a low or high signal) or can be set "tri-stated" or "high-Z" and effectively disconnected (not driven high or low). When the output enable signal (the one at the bottom of the triangle) is low, the output signal is driven. When it's high, the output is disconnected. The circle means it's an active-low or inverted signal.

The "PWM" signal is passed through, and raised to a higher voltage. It is always driven.

The "A̅T̅T̅E̅N̅U̅A̅T̅E̅" (the overline means an active low signal) signal causes the output to be pulled low (attenuate the voltage across the inductor) when pulled low. I haven't read through all of your code, so you may need to invert the attenuate signal in software to attenuate when you need the signal attenuated.

The inductor is the "antenna".

Something like a 74LVC125 should be suitable for this level shifting. It's a 4-buffer device, so tie the unused inputs high or low (connect the unused inputs to either power or ground.) It should be powered by a 3.3V supply.

<img width="608" alt="Screen Shot 2022-02-15 at 7 50 29 AM" src="https://user-images.githubusercontent.com/1057065/154065499-1ca215dc-9225-4856-8d50-9857b276be7a.png">

icb- commented:
> So a buffer is something like a transistor, right?

Not quite.

A buffer like in the 74125 can drive the output high, low, or high impedance (not driven, tri-state, high-Z). It's a digital device.

A BJT like you have in your circuit works kind of like a switch. It's not quite a switch, it's a current amplifier, but that's a good enough analogy for now. You should use a PNP BJT on the high side instead of the NPN you have in your diagram.

That might be good enough for how this circuit works.

You're not going to get away with less than two transistors. One is switching the higher voltage across the "antenna" from the PWM signal. The other is switching in a resistor in parallel to the "antenna" to attenuate the signal.

It might help to redraw the circuit more like this to show the two parallel current paths, the "antenna" and the attenuator.

<img width="228" alt="Screen Shot 2022-02-15 at 1 35 28 PM" src="https://user-images.githubusercontent.com/1057065/154126856-85b44a12-5214-4942-badf-e1a02e1ba7f1.png">

I say "antenna" in quotes and draw it as an inductor, because I don't think this is really acting as a radio transmitter, but is inductively coupling the signal into the antenna on the clock.

pjueon commented:
> You should use a PNP BJT on the high side instead of the NPN you have in your diagram.

Why? In the moment I only have NPN transistors. And I thought it would be no difference. Is there any difference in this application?

pjueon commented:
> If the voltage is 1.8V, this indidcates that the GPIO ports are connected to the internal core voltage of the CPU and they will have issue to drive anything and it will probably dangerous to even connecting the coil, as these tiny mosfets in the output stage will not be able to move the current.
> 
> I strongly suggest to connect the GPIO pins via a 1.8V -> 3.3V level shifter to (a) relieve the GPIO pins (b) have enough drive strength for the external components such as coil and pulldown.

After reading the following document many times( and thanks to icb-'s  explanation) ,I think I'm just starting to understand it.

[customizing_the_jetson_nano_40-pin_expansion_header_v1.2.pdf.pdf](https://github.com/hzeller/txtempus/files/8068264/customizing_the_jetson_nano_40-pin_expansion_header_v1.2.pdf.pdf)

According to this document, the Jetson Nano's GPIO pin headers are already connected to the level shifters (TXB0108).
 
![image](https://user-images.githubusercontent.com/52922372/154262089-d72c2b8a-cd5c-4556-82f4-a7495a4e49ae.png)

When I set the GPIO output to HIGH, the pin is being pulled up to 3.3v by 4k ohm internally.
And when I set it to LOW, it is being pulled down to the GND with 4k ohm internally. 


When I said the output pin voltage was ~1.5V, I connected 4~5k ohm to the pin. 
If I connected ~25k ohm to the pin and set the pin HIGH (without PWM) I got ~3V. 

From the document:
> The TXB0108 devices are not intended for use in open-drain applications or to drive low impedance (i.e. directly driving an LED) or high capacitive loads (i.e. driving a long wire, or multiple loads).

So maybe I should use more registers.  The reason why I couldn't pull down the signal  to the GND even if I set the pin to LOW from my first setup, it's probably because the pin is internally pulled down to  GND with 4k ohm. I'll give a try again with higher registers !

icb- commented:
>> You should use a PNP BJT on the high side instead of the NPN you have in your diagram.

> Why?

That's just how transistors work. To fully turn an NPN BJT on (saturate), the base needs to be at a higher voltage than the collector. If you have it on the high side of the circuit, you can't ever reach that. There are lots of explanations of why on the Internet, like [this one](https://www.baldengineer.com/low-side-vs-high-side-transistor-switch.html) and [this one](https://web.engr.oregonstate.edu/~traylor/ece112/beamer_lectures/bjt_reg_of_op.pdf) and the linked resources from them.

Technically, you should also have a resistor between the control pins and the base of each transistor to limit the base current, but it looks like your level shifter is already doing that for you.

If you need to use only NPN transistors, it would be better to drive it from the low side like this.

<img width="224" alt="Screen Shot 2022-02-16 at 7 35 13 AM" src="https://user-images.githubusercontent.com/1057065/154265555-2cc24eca-98af-42ac-b681-9298cb71a3b7.png">

> The TXB0108 devices are not intended for use in open-drain applications

This is why you can't drive the attenuate signal directly from the level shifter. You need to be able to drive it low, but not high. That's what open-drain is (or open-collector with BJTs). You may be able to drive the PWM signal directly from the level shifter, and the ATTENUATE signal with a transistor like this, if the level shifter can source enough current.

<img width="225" alt="Screen Shot 2022-02-16 at 7 59 23 AM" src="https://user-images.githubusercontent.com/1057065/154269576-49b55241-9d2d-4c1a-8c3e-d1611c1838a4.png">

pjueon commented:
@icb-  I just tested the second image you posted, and it DOES work!

<img alt="Screen Shot 2022-02-16 at 7 59 23 AM" width="225" src="https://user-images.githubusercontent.com/1057065/154269576-49b55241-9d2d-4c1a-8c3e-d1611c1838a4.png">

> That's just how transistors work. To fully turn an NPN BJT on (saturate), the base needs to be at a higher voltage than the collector. If you have it on the high side of the circuit, you can't ever reach that.

I see! I really didn't know that. Thanks again for your kind explanation. 
I truly had zero knowledge about this kind of stuff, haha.


Please forgive me one more basic question.
I found this setup from this video:  https://youtu.be/6SHGAEhnsYk?t=596

![image](https://user-images.githubusercontent.com/52922372/154281002-6eed2e78-66b1-4845-a1f3-bcdbcef36834.png)

This setup looks like switching the signal instead of attenuating it. Is my understanding correct?

icb- commented:
> This setup looks like switching the signal instead of attenuating it. Is my understanding correct?

It's doing both. It's operating the transistor in its active region, where it's amplifying the base-emitter current.

When GPIO4 is high, a small current flows into the base, resulting in a large current flowing through the transistor collector-emitter. This is because R3 is a relatively high value, so a relatively smaller current flows into the base.

When GPIO17 is high, a larger current flows into the base resulting in an even larger current flowing through the transistor collector-emitter. This is because R2 is a relatively low value, so a relatively larger current flows into the base.

When both GPIO4 and GPIO17 are high, you have a still larger current.

pjueon commented:
@hzeller 
I've just pushed some commits to this branch. 
Could you check if I properly fixed everything I need to?

pjueon commented:
Oops, I almost forgot to update the .gitignore file. Fixed it now!

JDat commented:
Pardon my electronics!
1st schematic (two transistors in series) doesn't looks good, because if lower transistor is closed, upper will get negative bias on  base relative to emitter.

Las schematic (different GPIOs on one transistor base) looks fine, but need some fine tuning with base resistors. Not for beginners.

What about  this schematic? More components, but... Fine tune R6 (set and forget + share to community) and that's it!
PS: Even if I like R2 and R4, they can be omitted.

![2022-02-16-194657_1366x744_scrot](https://user-images.githubusercontent.com/3792125/154325674-db4b3520-2bb4-48ed-aa81-ea2153138d1d.png)
.

pjueon commented:
> Pardon my electronics! 1st schematic (two transistors in series) doesn't looks good, because if lower transistor is closed, upper will get negative bias on base relative to emitter.
> 
> Las schematic (different GPIOs on one transistor base) looks fine, but need some fine tuning with base resistors. Not for beginners.
> 
> What about this schematic? More components, but... Fine tune R6 (set and forget + share to community) and that's it! PS: Even if I like R2 and R4, they can be omitted.
> 
> ![2022-02-16-194657_1366x744_scrot](https://user-images.githubusercontent.com/3792125/154325674-db4b3520-2bb4-48ed-aa81-ea2153138d1d.png) .

Thanks for the advice!
Hmm, then how do you think about this one? Personally I like it because it only requires 1 transistor and looks relatively simple for me.

<img alt="Screen Shot 2022-02-16 at 7 59 23 AM" width="225" src="https://user-images.githubusercontent.com/1057065/154269576-49b55241-9d2d-4c1a-8c3e-d1611c1838a4.png">

I tested this setup and it works well. So if you think this setup has no big issue, I think it would be good enough for this project.

pjueon commented:
<Reminder>
@hzeller Could you check this please? 

What I fixed:
- added new line at the end of the files
- changed the setup
  - the level shifter was not needed because it was already included in the Jetson devices 
  - now it modulates the signal by attenuating the pwm signal instead of turning on/off 
- changed README
- changed the images 
- fixed macro names

pjueon commented:
Reminder again.
@hzeller 

Could you check this please...?

hzeller reviewed (approved):
Thanks!
Sorry for the delay, sometimes notifications from github vanish in the noise of email :)
